pkg/meta/boltdb.go

```diff
@@ -0,0 +1,140 @@
+package meta
+
+import (
+	"os"
+	"path"
+	"time"
+
+	boltdb "github.com/boltdb/bolt"
+	"github.com/pkg/errors"
+)
+
+func init() {
+	Register("boltdb", &bolt{})
+}
+
+type bolt struct {
+	db *boltdb.DB
+}
+
+// New is used to make bolt metadata store instance.
+func (b *bolt) New(cfg Config) error {
+	opt := &boltdb.Options{
+		Timeout: time.Second * 10,
+	}
+
+	if _, err := os.Stat(cfg.BaseDir); err != nil && os.IsNotExist(err) {
+		if err := os.MkdirAll(path.Dir(cfg.BaseDir), 0755); err != nil {
+			return errors.Wrap(err, "failed to create metadata path")
+		}
+	}
+
+	db, err := boltdb.Open(cfg.BaseDir, 0644, opt)
+	if err != nil {
+		return err
+	}
+	for _, bucket := range cfg.Buckets {
+		if err := b.prepare(db, []byte(bucket.Name)); err != nil {
+			return err
+		}
+	}
+	b.db = db
+
+	return nil
+}
+
+func (b *bolt) prepare(db *boltdb.DB, bucket []byte) error {
+	return db.Update(func(tx *boltdb.Tx) error {
+		_, err := tx.CreateBucketIfNotExists(bucket)
+		if err != nil {
+			return errors.Wrap(err, "boltdb: create bucket")
+		}
+		return nil
+	})
+}
+
+// Path returns boltdb store file.
+func (b *bolt) Path(key string) string {
+	return b.db.Path()
+}
+
+// Keys return all keys for boltdb.
+func (b *bolt) Keys(bucket string) ([]string, error) {
+	keys := make([]string, 0)
+
+	err := b.db.View(func(tx *boltdb.Tx) error {
+		bkt := tx.Bucket([]byte(bucket))
+		if bkt == nil {
+			return ErrBucketNotFound
+		}
+
+		return bkt.ForEach(func(k, v []byte) error {
+			keys = append(keys, string(k))
+			return nil
+		})
+	})
+
+	return keys, err
+}
+
+// Put is used to put metadate into boltdb.
+func (b *bolt) Put(bucket, key string, value []byte) error {
+	return b.db.Update(func(tx *boltdb.Tx) error {
+		bkt := tx.Bucket([]byte(bucket))
+		if bkt == nil {
+			return ErrBucketNotFound
+		}
+		if err := bkt.Put([]byte(key), value); err != nil {
+			return errors.Wrap(err, "boltdb: put key: "+key)
```

Comment:
Two much colon, the result error string would be `boltdb: put key: sdfgh: err.Error()`, how about `failed to put key asdfgh in boltdb: err.Error()` using `errors.Wrap(err, fmt.Sprintf("failed to put key %s in boltdb", key))`